src/horizontal-pod-autoscaler.ts

```diff
@@ -0,0 +1,865 @@
+   */
+  public readonly metrics?: Metric[];
+  /**
+   * The scaling behavior when scaling up.
+   *
+   * @default
+   * Is the higher of:
+   * * Increase no more than 4 pods per 60 seconds
+   * * Double the number of pods per 60 seconds
+   */
+  public scaleUp: IScalingRules;
+  /**
+   * The scaling behavior when scaling down.
+   *
+   * @default
+   * Scale down to minReplica count with a 5 minute stabilization window.
+   */
+  public scaleDown: IScalingRules;
+
+  // Defaults
+  private readonly _defaultScaleUp: IScalingRules = {
+    strategy: ScalingStrategy.MAX_CHANGE,
+    stabilizationWindow: Duration.seconds(0),
+    policies: [
+      {
+        replicas: Replicas.absolute(4),
+        duration: Duration.minutes(1),
+      },
+      {
+        replicas: Replicas.percent(200),
+        duration: Duration.minutes(1),
+      },
+    ],
+  };
+  private readonly _defaultScaleDown: IScalingRules = {
+    strategy: ScalingStrategy.MAX_CHANGE,
+    stabilizationWindow: Duration.minutes(5),
+    policies: [
+      {
+        replicas: Replicas.absolute(this.minReplicas ?? 1),
+        duration: Duration.minutes(5),
+      },
+    ],
+  };
+  private readonly _defaultScalingPolicy = {
+    duration: Duration.seconds(15),
+  };
+
+
+  constructor(scope: Construct, id: string, props: HorizontalPodAutoscalerProps) {
+    super(scope, id);
+    this.apiObject = new k8s.KubeHorizontalPodAutoscalerV2(this, 'Resource', {
+      metadata: props.metadata,
+      spec: Lazy.any({ produce: () => this._toKube() }),
+    });
+
+    if (props.target.kind === 'DaemonSet') {
+      throw new Error('HorizontalPodAutoscaler cannot be used with workloads that do not scale, such as a DaemonSet.');
+    }
+    const hasResourceConstraints = props.target.containers.some((c) => this._hasRequestsOrLimits(c));
+    if (!hasResourceConstraints && !props.metrics) {
+      throw new Error('Every container in the HorizontalPodAutoscaler target must have CPU or memory resources defined');
+    }
+    if (props?.minReplicas && props.minReplicas > props.maxReplicas) {
+      throw new Error("'minReplicas' must be less than or equal to maxReplicas in order for HorizontalPodAutoscaler to scale.");
+    }
+    if (props?.scaleUp?.stabilizationWindow && !this._isStabilizationWindowWithinRange(props.scaleUp.stabilizationWindow)) {
+      throw new Error("'scaleUp.stabilizationWindow' must be more than 0 seconds and no longer than 1 hour.");
+    }
+    if (props?.scaleDown?.stabilizationWindow && !this._isStabilizationWindowWithinRange(props.scaleDown.stabilizationWindow)) {
+      throw new Error("'scaleDown.stabilizationWindow' must be more than 0 seconds and no longer than 1 hour.");
+    }
+    if (props?.scaleUp?.policies?.length) {
+      this._validateScalingPolicies(ScalingDirection.UP, props.scaleUp.policies);
+      props.scaleUp.policies = this._setDefaultsOnScalingPolicies(props.scaleUp.policies);
```

Comment:
Yeah It felt gross doing it. I was having some struggles with making JSII and TS happy. Was also trying to avoid setting the default values in the `_toKube()`. I'll rework it.